adapters/repos/db/vector/hnsw/insert.go

```diff
@@ -151,18 +151,39 @@ func (h *hnsw) insert(node *vertex, nodeVec []float32) error {
 
 	nodeId := node.id
 
-	// before = time.Now()
-	h.Lock()
-	// m.addBuildingLocking(before)
-	err := h.growIndexToAccomodateNode(node.id, h.logger)
-	if err != nil {
+	// There are 3 different things going on in this block:
+	// 1) A global read lock to check if we need to grow the index and only upgrade to a RW lock if yes
+	// 2) The write to the nodes-slice needs to be behind a Rlock to secure it against growing from another thread. A
+	//    RW-lock is not necessary as the node is not discoverable until the connections with the neighboring nodes have
+	//	  been established.
+	// 3) The sharded-lock secures against searches with an allowList. An object can already be added to the object
+	//    store before this insert completes and cause a race between writing to the nodes-slice and reading all nodes
+	//	  from the allowlist.
```

Comment:
not sure I understand this one, why is the lock situation different if an allow list is present?